scrape/target_test.go

```diff
@@ -354,6 +354,80 @@ func TestTargetsFromGroup(t *testing.T) {
 	require.EqualError(t, failures[0], expectedError)
 }
 
+// TestTargetsFromGroupWithLabelKeepDrop aims to demonstrate and reinforce the current behavior: relabeling's "labelkeep" and "labeldrop"
+// are applied to all labels of a target, including internal ones (labels starting with "__" such as "__address__").
+// This will be hlpful for cases like https://github.com/prometheus/prometheus/issues/12355.
+func TestTargetsFromGroupWithLabelKeepDrop(t *testing.T) {
+	tests := []struct {
+		name             string
+		cfgText          string
+		targets          []model.LabelSet
+		shouldDropTarget bool
+	}{
+		{
+			name: "no relabeling",
+			cfgText: `
+global:
+  metric_name_validation_scheme: legacy
```

Comment:
Had to add those to get around:

```
2024/12/03 17:19:06 http: TLS handshake error from 127.0.0.1:52972: remote error: tls: bad certificate
--- FAIL: TestTargetsFromGroupWithLabelKeepDrop (0.00s)
    --- FAIL: TestTargetsFromGroupWithLabelKeepDrop/no_relabeling (0.00s)
panic: utf8 name validation requested but model.NameValidationScheme is not set to UTF8 [recovered]
	panic: utf8 name validation requested but model.NameValidationScheme is not set to UTF8 [recovered]
	panic: utf8 name validation requested but model.NameValidationScheme is not set to UTF8
```

(will look into that later, maybe we need to override a global var in this test file or sth...)

Reply:
opened https://github.com/prometheus/prometheus/issues/15530 as this is not caused/related to the change in the PR.

Reply:
Thanks for opening #15530 .